syntax = "proto3";
import "parameter.proto";
import "indexer.proto";

package retrieve;

/**
    Service to process and retrieve documents
*/
service Retriever {
    rpc GetRetrievers(GetRetrieversRequest) returns (GetRetrieversResponse);
    rpc Retrieve(RetrieveRequest) returns (RetrieveResponse);
}

message RetrieverComponent {
    string retriever_id = 1;
    repeated parameter.Parameter parameters = 2;
    string engine_type = 3;
}

message GetRetrieversRequest {
}

message GetRetrieversResponse {
    repeated RetrieverComponent retrievers = 1;
}


message RetrieveRequest {
    RetrieverComponent retriever = 1;
    string index_id = 2; 
    repeated string queries = 3;
}

message Hit {
    index.Document document = 1;
    double score = 2;
}

message HitPerQuery {
    repeated Hit hits = 1;
}

message RetrieveResponse {
    repeated HitPerQuery hits = 1;
}
